File: 3639-zero-array-transformation-i/3639-zero-array-transformation-i.cpp
class Solution {
public:
    bool isZeroArray(vector<int>& nums, vector<vector<int>>& queries) {
        int n = nums.size();
        vector<int> prefix (n+1, 0);
        
        for(vector<int>& q: queries){
            int left = q[0];
            int right = q[1];
            prefix[left]++;
            prefix[right+1]--; 
        }

        for(int i = 1; i < n; i++){
            prefix[i] += prefix[i-1];
        }

        for(int i = 0; i < n; i++){
            nums[i] -= prefix[i];
            if(nums[i] > 0) return false;
        }

        return true;
    }
};

File: 3639-zero-array-transformation-i/3639-zero-array-transformation-i.py
class Solution(object):
    def isZeroArray(self, nums, queries):
        """
        :type nums: List[int]
        :type queries: List[List[int]]
        :rtype: bool
        """
        n = len(nums)
        prefix = [0] * (n+1)

        for q in queries:
            left, right = q[0], q[1]
            prefix[left] += 1
            prefix[right+1] -= 1

        for i in range(1, n+1):
            prefix[i] += prefix[i-1]
        for i in range(n):
            # print(prefix[i])
            nums[i] -= prefix[i]
            if(nums[i] > 0):
                return False
        return True
        